rules/otp.bzl

```diff
@@ -0,0 +1,233 @@
+                # Cryptolib and chip-level tests require access to the
+                # entropy_src FW data interface.
+                "EN_ENTROPY_SRC_FW_READ": True,
+                # Cryptolib and chip-level tests require access to the
+                # entropy_src FW override interface.
+                "EN_ENTROPY_SRC_FW_OVER": True,
+            },
+        },
+        {
+            "name": "SECRET0",
+            "lock": True,
+            "items": {
+                "TEST_UNLOCK_TOKEN": "<random>",
+                "TEST_EXIT_TOKEN": "<random>",
+            },
+        },
+        {
+            "name": "SECRET1",
+            "lock": True,
+            "items": {
+                "FLASH_ADDR_KEY_SEED": "<random>",
+                "FLASH_DATA_KEY_SEED": "<random>",
+                "SRAM_DATA_KEY_SEED": "<random>",
+            },
+        },
+        {
+            "name": "SECRET2",
+            "lock": False,
+            "items": {
+                "RMA_TOKEN": "<random>",
+                "CREATOR_ROOT_KEY_SHARE0": "<random>",
+                "CREATOR_ROOT_KEY_SHARE1": "<random>",
+            },
+        },
+        {
+            "name": "LIFE_CYCLE",
+            "state": ctx.attr.lc_state,
+            # Can range from 0 to 16.
+            # Note that a value of 0 is only permissible in RAW state.
+            "count": ctx.attr.lc_count,
+        },
+    ]
+
+    # For every partition with an "items" dictionary, expand the dictionary of
+    # key:value pairs into a list of dicts, each of the form
+    # {
+    #   "name": key,
+    #   "value": value
+    # }
+    # This format is expected by the OTP image generation tool
+    for partition in otp["partitions"]:
+        if "items" in partition.keys():
+            items = partition["items"]
+            partition["items"] = [{"name": k, "value": items[k]} for k in items.keys()]
+
+    file = ctx.actions.declare_file("{}.json".format(ctx.attr.name))
+    ctx.actions.write(file, json.encode_indent(otp))
+    return DefaultInfo(files = depset([file]))
+
+otp_json = rule(
+    implementation = _otp_json_impl,
+    attrs = {
+        # Valid life cycle states can be found in the life cycle state
+        # definition file (default: hw/ip/lc_ctrl/data/lc_ctrl_state.hjson)
+        "lc_state": attr.string(doc = "Life cycle state", default = "RMA"),
+        "lc_count": attr.int(doc = "Life cycle count", default = 8),
+    },
+)
+
+def _otp_image(ctx):
+    # TODO(dmcardle) I don't like hardcoding the width in the filename. Maybe we
+    # can write it into some metadata in the file instead.
+    output = ctx.actions.declare_file(ctx.attr.name + ".24.vmem")
+    args = ctx.actions.args()
+    args.add("--quiet")
+    args.add("--lc-state-def", ctx.file.lc_state_def)
+    args.add("--mmap-def", ctx.file.mmap_def)
+    args.add("--img-cfg", ctx.file.src)
+    args.add("--out", "{}/{}.BITWIDTH.vmem".format(output.dirname, ctx.attr.name))
```

Comment:
nit: `BITWIDTH`?

Reply:
Hey Alphan, could you expand on this nit? The "BITWIDTH" substring gets replaced in `gen-otp-img.py`[^0] with an integer (in this case, 24). The Python class `OtpMemImg` determines the width over here[^1].

@jon-flatley FYI in case you were not aware of this functionality for the Rust implementation :)

[^0]: https://cs.opensource.google/opentitan/opentitan/+/master:util/design/gen-otp-img.py;l=270;drc=3da5eab1ed884a516c35cca5db182e0ad0acb024
[^1]: https://cs.opensource.google/opentitan/opentitan/+/master:util/design/lib/OtpMemImg.py;l=100;drc=3da5eab1ed884a516c35cca5db182e0ad0acb024